### PR TITLE
Dockerregistry config refactor

### PR DESCRIPTION
**Pull request by legionus**

### CODE REVIEW

File: pkg/dockerregistry/server/configuration/configuration.go [hunk 1]
```diff
@@ -3,18 +3,79 @@ package configuration
...
 	"io/ioutil"
 	"os"
 	"reflect"
 	"strings"
 	"time"
 
+	log "github.com/Sirupsen/logrus"
 	"gopkg.in/yaml.v2"
 
+	//"github.com/docker/distribution/registry/auth"
```
Comment (dmage) [inline pkg/dockerregistry/server/configuration/configuration.go]: remove it

File: vendor/github.com/docker/distribution/configuration/parser.go [hunk 2]
```diff
@@ -220,7 +220,7 @@ func (p *Parser) overwriteStruct(v reflect.Value, fullpath string, path []string
 		}
 	case reflect.Ptr:
 		if field.IsNil() {
-			field.Set(reflect.New(sf.Type))
```
Comment (dmage) [inline vendor/github.com/docker/distribution/configuration/parser.go]: sf.Type.Elem()?

File: pkg/dockerregistry/server/blobdescriptorservice_test.go [hunk 3]
```diff
@@ -78,7 +77,17 @@ func TestBlobDescriptorServiceIsApplied(t *testing.T) {
 			"repository": {{Name: "openshift"}},
 			"storage":    {{Name: "openshift"}},
 		},
-	}, &srvconfig.Configuration{}, nil)
+	}
+
+	cfg := &srvconfig.Configuration{
+		Server: &srvconfig.Server{":5000"},
+	}
+	if err := srvconfig.InitExtraConfig(dockercfg, cfg); err != nil {
```
Comment (dmage) [inline pkg/dockerregistry/server/blobdescriptorservice_test.go]: Don't you want to move it into NewApp?

File: pkg/dockerregistry/server/blobdescriptorservice_test.go [hunk 4]
```diff
@@ -78,7 +77,17 @@ func TestBlobDescriptorServiceIsApplied(t *testing.T) {
...
+		t.Fatal(err)
+	}
+
+	os.Setenv("OPENSHIFT_DEFAULT_REGISTRY", "localhost:5000")
```
Comment (dmage) [inline pkg/dockerregistry/server/blobdescriptorservice_test.go]: Use config instead.

File: images/dockerregistry/config.yml [hunk 5]
```diff
@@ -58,3 +44,13 @@ openshift:
       maxrunning: 0
       maxinqueue: 0
       maxwaitinqueue: 0
+  quota:
+    enabled: false
+  cache:
+    blobrepositoryttl: 10m
+    quotattl: 1m
```
Comment (dmage) [inline images/dockerregistry/config.yml]: Move quotattl to the quota section.
/cc @miminar

File: pkg/dockerregistry/server/quotarestrictedblobstore.go [hunk 6]
```diff
@@ -32,33 +33,23 @@ const (
...
 // timeout. Caches will only be initialized if the given ttl is positive. Options are gathered from
 // configuration file and will be overridden by enforceQuota and projectCacheTTL environment variable values.
-func newQuotaEnforcingConfig(ctx context.Context, enforceQuota, projectCacheTTL string, options map[string]interface{}) *quotaEnforcingConfig {
-	enforce, err := getBoolOption(EnforceQuotaEnvVar, "enforcequota", false, options)
-	if err != nil {
-		context.GetLogger(ctx).Error(err)
-	}
-
-	if !enforce {
+func newQuotaEnforcingConfig(ctx context.Context, cfg *configuration.Configuration) *quotaEnforcingConfig {
```
Comment (dmage) [inline pkg/dockerregistry/server/quotarestrictedblobstore.go]: s/configuration.Configuration/configuration.Quota/

File: pkg/dockerregistry/server/configuration/configuration.go (hunk 1 above)
Comment (dmage) [reply pkg/dockerregistry/server/configuration/configuration.go]: This import is for the commented section below.

File: vendor/github.com/docker/distribution/configuration/parser.go (hunk 2 above)
Comment (dmage) [reply vendor/github.com/docker/distribution/configuration/parser.go]: We already have the upstream PR.

File: pkg/dockerregistry/server/blobdescriptorservice_test.go (hunk 3 above)
Comment (dmage) [reply pkg/dockerregistry/server/blobdescriptorservice_test.go]: Start() wants the initialised config as well.

File: pkg/dockerregistry/server/auth_test.go [hunk 7]
```diff
@@ -425,9 +431,14 @@ func TestAccessController(t *testing.T) {
 		app := &App{
 			ctx:            ctx,
 			registryClient: client.NewRegistryClient(cfg),
-			extraConfig:    &configuration.Configuration{},
+			extraConfig: &configuration.Configuration{
+				Server: &configuration.Server{":5000"},
```
Comment (dmage) [inline pkg/dockerregistry/server/auth_test.go]: add Addr:

Comment (dmage) [reply pkg/dockerregistry/server/auth_test.go]: s/:5000/localhost:5000/

Comment (legionus) [reply pkg/dockerregistry/server/auth_test.go]: @dmage done

File: pkg/dockerregistry/server/quotarestrictedblobstore.go (hunk 6 above)
Comment (legionus) [reply pkg/dockerregistry/server/quotarestrictedblobstore.go]: @dmage done

File: images/dockerregistry/config.yml (hunk 5 above)
Comment (legionus) [reply images/dockerregistry/config.yml]: @dmage @miminar done

**Pull request merged**